File: native_modules/camera_native/ios/Classes/CameraNativePlugin.swift
import Flutter
import UIKit
import AVFoundation
import CoreImage
import CoreML
import Vision
import Photos

// External XCFramework imports
// import AdvancedCameraSDK
// import AIImageProcessor
// import CameraMLKit

public class CameraNativePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var channel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    // Camera components
    private var captureSession: AVCaptureSession?
    private var photoOutput: AVCapturePhotoOutput?
    private var videoOutput: AVCaptureMovieFileOutput?
    private var currentCamera: AVCaptureDevice?

    // External XCFramework instances
    private var advancedCameraSDK: Any? // AdvancedCameraSDK instance
    private var aiImageProcessor: Any? // AIImageProcessor instance
    private var cameraMLKit: Any? // CameraMLKit instance

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "camera_native", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "camera_native_events", binaryMessenger: registrar.messenger())

        let instance = CameraNativePlugin()
        instance.channel = channel
        instance.eventChannel = eventChannel

        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    public override init() {
        super.init()
        initializeExternalFrameworks()
    }

    private func initializeExternalFrameworks() {
        DispatchQueue.global(qos: .background).async { [weak self] in
            do {
                // Initialize AdvancedCameraSDK XCFramework
                // self?.advancedCameraSDK = try AdvancedCameraSDK.shared.initialize()

                // Initialize AIImageProcessor XCFramework
                // self?.aiImageProcessor = try AIImageProcessor.create()

                // Initialize CameraMLKit XCFramework
                // self?.cameraMLKit = try CameraMLKit.shared.setup()

                DispatchQueue.main.async {
                    self?.sendEvent(data: [
                        "type": "external_frameworks_initialized",
                        "frameworks": ["AdvancedCameraSDK", "AIImageProcessor", "CameraMLKit"]
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    self?.sendEvent(data: [
                        "type": "error",
                        "message": "Failed to initialize XCFrameworks: \(error.localizedDescription)"
                    ])
                }
            }
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "initializeCamera":
            initializeCamera(result: result)

        case "capturePhoto":
            capturePhoto(result: result)

        case "startVideoRecording":
            startVideoRecording(result: result)

        case "stopVideoRecording":
            stopVideoRecording(result: result)

        case "enableAdvancedFeatures":
            if let arguments = call.arguments as? [String: Any] {
                enableAdvancedFeatures(arguments: arguments, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            }

        case "processImageWithAI":
            if let arguments = call.arguments as? [String: Any],
               let imagePath = arguments["imagePath"] as? String {
                processImageWithAI(imagePath: imagePath, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Image path required", details: nil))
            }

        case "detectObjectsInRealTime":
            detectObjectsInRealTime(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initializeCamera(result: @escaping FlutterResult) {
        guard hasPermissions() else {
            result(FlutterError(code: "PERMISSION_DENIED", message: "Camera permissions not granted", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                let session = AVCaptureSession()
                session.sessionPreset = .photo

                // Configure camera input
                guard let camera = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back) else {
                    DispatchQueue.main.async {
                        result(FlutterError(code: "NO_CAMERA", message: "No camera available", details: nil))
                    }
                    return
                }

                let input = try AVCaptureDeviceInput(device: camera)
                session.addInput(input)

                // Configure photo output
                let photoOutput = AVCapturePhotoOutput()
                session.addOutput(photoOutput)

                // Configure video output
                let videoOutput = AVCaptureMovieFileOutput()
                session.addOutput(videoOutput)

                self?.captureSession = session
                self?.photoOutput = photoOutput
                self?.videoOutput = videoOutput
                self?.currentCamera = camera

                session.startRunning()

                DispatchQueue.main.async {
                    self?.sendEvent(data: ["type": "camera_initialized"])
                    result(true)
                }
            } catch {
                DispatchQueue.main.async {
                    self?.sendEvent(data: [
                        "type": "error",
                        "message": "Camera initialization failed: \(error.localizedDescription)"
                    ])
                    result(FlutterError(code: "INIT_FAILED", message: "Camera initialization failed", details: error.localizedDescription))
                }
            }
        }
    }

    private func capturePhoto(result: @escaping FlutterResult) {
        guard let photoOutput = photoOutput else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }

        let settings = AVCapturePhotoSettings()
        settings.flashMode = .auto

        // Use advanced camera features from XCFramework
        // if let advancedSDK = advancedCameraSDK as? AdvancedCameraSDK {
        //     settings = advancedSDK.enhancePhotoSettings(settings)
        // }

        let delegate = PhotoCaptureDelegate { [weak self] imagePath in
            if let path = imagePath {
                self?.sendEvent(data: ["type": "photo_captured", "path": path])
                result(path)
            } else {
                result(FlutterError(code: "CAPTURE_FAILED", message: "Photo capture failed", details: nil))
            }
        }

        photoOutput.capturePhoto(with: settings, delegate: delegate)
    }

    private func enableAdvancedFeatures(arguments: [String: Any], result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                let hdrEnabled = arguments["hdr"] as? Bool ?? false
                let nightModeEnabled = arguments["nightMode"] as? Bool ?? false
                let portraitModeEnabled = arguments["portraitMode"] as? Bool ?? false

                // Configure advanced features using XCFramework
                // if let advancedSDK = self?.advancedCameraSDK as? AdvancedCameraSDK {
                //     try advancedSDK.enableHDR(hdrEnabled)
                //     try advancedSDK.enableNightMode(nightModeEnabled)
                //     try advancedSDK.enablePortraitMode(portraitModeEnabled)
                // }

                DispatchQueue.main.async {
                    self?.sendEvent(data: [
                        "type": "advanced_features_configured",
                        "hdr": hdrEnabled,
                        "nightMode": nightModeEnabled,
                        "portraitMode": portraitModeEnabled
                    ])
                    result(true)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "FEATURE_ERROR", message: "Failed to enable advanced features", details: error.localizedDescription))
                }
            }
        }
    }

    private func processImageWithAI(imagePath: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                guard let image = UIImage(contentsOfFile: imagePath) else {
                    DispatchQueue.main.async {
                        result(FlutterError(code: "INVALID_IMAGE", message: "Cannot load image", details: nil))
                    }
                    return
                }

                // Process image using AI XCFramework
                // if let aiProcessor = self?.aiImageProcessor as? AIImageProcessor {
                //     let enhancedImage = try aiProcessor.enhance(image)
                //     let enhancedPath = imagePath.replacingOccurrences(of: ".jpg", with: "_enhanced.jpg")
                //
                //     if let data = enhancedImage.jpegData(compressionQuality: 0.9) {
                //         try data.write(to: URL(fileURLWithPath: enhancedPath))
                //
                //         DispatchQueue.main.async {
                //             self?.sendEvent(data: [
                //                 "type": "ai_processing_complete",
                //                 "originalPath": imagePath,
                //                 "enhancedPath": enhancedPath
                //             ])
                //             result(enhancedPath)
                //         }
                //     }
                // }

                // Simulate AI processing for now
                let enhancedPath = imagePath.replacingOccurrences(of: ".jpg", with: "_enhanced.jpg")
                DispatchQueue.main.async {
                    self?.sendEvent(data: [
                        "type": "ai_processing_complete",
                        "originalPath": imagePath,
                        "enhancedPath": enhancedPath
                    ])
                    result(enhancedPath)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "AI_ERROR", message: "AI processing failed", details: error.localizedDescription))
                }
            }
        }
    }

    private func detectObjectsInRealTime(result: @escaping FlutterResult) {
        // Use CameraMLKit XCFramework for real-time object detection
        // if let mlKit = cameraMLKit as? CameraMLKit {
        //     mlKit.startObjectDetection { [weak self] objects in
        //         self?.sendEvent(data: [
        //             "type": "objects_detected",
        //             "objects": objects
        //         ])
        //     }
        // }

        result(true)
    }

    private func startVideoRecording(result: @escaping FlutterResult) {
        // Video recording implementation
        result(true)
    }

    private func stopVideoRecording(result: @escaping FlutterResult) {
        // Stop video recording implementation
        result("/path/to/video.mp4")
    }

    private func hasPermissions() -> Bool {
        return AVCaptureDevice.authorizationStatus(for: .video) == .authorized
    }

    private func sendEvent(data: [String: Any]) {
        eventSink?(data)
    }

    // MARK: - FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

// MARK: - Photo Capture Delegate
class PhotoCaptureDelegate: NSObject, AVCapturePhotoCaptureDelegate {
    private let completion: (String?) -> Void

    init(completion: @escaping (String?) -> Void) {
        self.completion = completion
    }

    func photoOutput(_ output: AVCapturePhotoOutput, didFinishProcessingPhoto photo: AVCapturePhoto, error: Error?) {
        guard error == nil, let imageData = photo.fileDataRepresentation() else {
            completion(nil)
            return
        }

        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let fileName = "photo_\(Date().timeIntervalSince1970).jpg"
        let filePath = "\(documentsPath)/\(fileName)"

        do {
            try imageData.write(to: URL(fileURLWithPath: filePath))
            completion(filePath)
        } catch {
            completion(nil)
        }
    }
}

File: native_modules/camera_native/android/src/main/kotlin/com/superapp/camera_native/CameraNativePlugin.kt
package com.superapp.camera_native

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import androidx.lifecycle.LifecycleOwner
import java.io.File
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/** CameraNativePlugin with AAR dependencies integration */
class CameraNativePlugin: FlutterPlugin, MethodCallHandler, ActivityAware, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private lateinit var context: Context
    private var activityBinding: ActivityPluginBinding? = null

    // Camera components
    private var imageCapture: ImageCapture? = null
    private var videoCapture: VideoCapture? = null
    private var cameraProvider: ProcessCameraProvider? = null
    private lateinit var cameraExecutor: ExecutorService

    // External AAR libraries integration
    private var advancedCameraFeatures: Any? = null // From advanced-camera-features.aar
    private var aiImageProcessor: Any? = null // From ai-image-processing.aar

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "camera_native")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "camera_native_events")
        eventChannel.setStreamHandler(this)

        context = flutterPluginBinding.applicationContext
        cameraExecutor = Executors.newSingleThreadExecutor()

        // Initialize external AAR libraries
        initializeExternalLibraries()
    }

    private fun initializeExternalLibraries() {
        try {
            // Initialize advanced camera features from AAR
            // advancedCameraFeatures = AdvancedCameraSDK.initialize(context)

            // Initialize AI image processing from AAR
            // aiImageProcessor = AIImageProcessor.create(context)

            sendEvent(mapOf("type" to "external_libraries_initialized"))
        } catch (e: Exception) {
            sendEvent(mapOf("type" to "error", "message" to "Failed to initialize external libraries: ${e.message}"))
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "initializeCamera" -> {
                initializeCamera(result)
            }
            "capturePhoto" -> {
                capturePhoto(result)
            }
            "startVideoRecording" -> {
                startVideoRecording(result)
            }
            "stopVideoRecording" -> {
                stopVideoRecording(result)
            }
            "enableAdvancedFeatures" -> {
                enableAdvancedFeatures(call.arguments as Map<String, Any>, result)
            }
            "processImageWithAI" -> {
                processImageWithAI(call.argument<String>("imagePath")!!, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initializeCamera(result: Result) {
        if (!hasPermissions()) {
            result.error("PERMISSION_DENIED", "Camera permissions not granted", null)
            return
        }

        val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
        cameraProviderFuture.addListener({
            try {
                cameraProvider = cameraProviderFuture.get()

                // Build the image capture use case
                imageCapture = ImageCapture.Builder()
                    .setCaptureMode(ImageCapture.CAPTURE_MODE_MINIMIZE_LATENCY)
                    .build()

                // Build the video capture use case
                videoCapture = VideoCapture.Builder()
                    .setVideoFrameRate(30)
                    .build()

                sendEvent(mapOf("type" to "camera_initialized"))
                result.success(true)
            } catch (exc: Exception) {
                sendEvent(mapOf("type" to "error", "message" to "Camera initialization failed: ${exc.message}"))
                result.error("INIT_FAILED", "Camera initialization failed", exc.message)
            }
        }, ContextCompat.getMainExecutor(context))
    }

    private fun capturePhoto(result: Result) {
        val imageCapture = imageCapture ?: run {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }

        val name = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
        val contentValues = android.content.ContentValues().apply {
            put(android.provider.MediaStore.MediaColumns.DISPLAY_NAME, name)
            put(android.provider.MediaStore.MediaColumns.MIME_TYPE, "image/jpeg")
        }

        val outputOptions = ImageCapture.OutputFileOptions.Builder(
            context.contentResolver,
            android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            contentValues
        ).build()

        imageCapture.takePicture(
            outputOptions,
            ContextCompat.getMainExecutor(context),
            object : ImageCapture.OnImageSavedCallback {
                override fun onError(exception: ImageCaptureException) {
                    sendEvent(mapOf("type" to "error", "message" to "Photo capture failed: ${exception.message}"))
                    result.error("CAPTURE_FAILED", "Photo capture failed", exception.message)
                }

                override fun onImageSaved(output: ImageCapture.OutputFileResults) {
                    val savedUri = output.savedUri?.toString()
                    sendEvent(mapOf("type" to "photo_captured", "path" to savedUri))
                    result.success(savedUri)
                }
            }
        )
    }

    private fun enableAdvancedFeatures(arguments: Map<String, Any>, result: Result) {
        try {
            // Use advanced camera features from AAR
            val hdrEnabled = arguments["hdr"] as? Boolean ?: false
            val nightModeEnabled = arguments["nightMode"] as? Boolean ?: false

            // Example: Configure advanced features using external AAR
            // advancedCameraFeatures?.enableHDR(hdrEnabled)
            // advancedCameraFeatures?.enableNightMode(nightModeEnabled)

            sendEvent(mapOf(
                "type" to "advanced_features_configured",
                "hdr" to hdrEnabled,
                "nightMode" to nightModeEnabled
            ))
            result.success(true)
        } catch (e: Exception) {
            result.error("FEATURE_ERROR", "Failed to enable advanced features", e.message)
        }
    }

    private fun processImageWithAI(imagePath: String, result: Result) {
        try {
            // Use AI image processing from AAR
            // val processedImage = aiImageProcessor?.enhance(imagePath)

            // Simulate AI processing result
            val enhancedPath = imagePath.replace(".jpg", "_enhanced.jpg")

            sendEvent(mapOf(
                "type" to "ai_processing_complete",
                "originalPath" to imagePath,
                "enhancedPath" to enhancedPath
            ))
            result.success(enhancedPath)
        } catch (e: Exception) {
            result.error("AI_ERROR", "AI processing failed", e.message)
        }
    }

    private fun startVideoRecording(result: Result) {
        // Video recording implementation
        result.success(true)
    }

    private fun stopVideoRecording(result: Result) {
        // Stop video recording implementation
        result.success("/path/to/video.mp4")
    }

    private fun hasPermissions(): Boolean {
        return ContextCompat.checkSelfPermission(
            context, Manifest.permission.CAMERA
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun sendEvent(data: Map<String, Any?>) {
        eventSink?.success(data)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        cameraExecutor.shutdown()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onDetachedFromActivity() {
        activityBinding = null
    }
}
